### PR TITLE
outleirgram

### PR DESCRIPTION
**Pull request by mellamansanchez**

### CODE REVIEW

File: skfda/exploratory/visualization/_baseplot.py [hunk 1]
```diff
@@ -34,7 +34,7 @@ def __init__(
         fig: Optional[Figure] = None,
         axes: Union[Axes, Sequence[Axes], None] = None,
     ) -> None:
-        self.artists: np.ndarray
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_baseplot.py:37]: Why this change? The type is ndarray.

File: skfda/exploratory/visualization/_outliergram.py [hunk 2]
```diff
@@ -74,57 +103,45 @@ def plot(
...
-                arguments to be passed to the matplotlib.pyplot.plot_surface
-                function.
         Returns:
             fig: figure object in which the depths will be
             scattered.
         """
-        fig, axes_list = _get_figure_and_axes(chart, fig, axes)
-        fig, axes_list = _set_figure_layout_for_fdata(
-            self.fdata, fig, axes_list, n_rows, n_cols,
+        self.artists = np.array([])
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:110]: Create it with the final size.

File: skfda/exploratory/visualization/_outliergram.py [hunk 3]
```diff
@@ -74,57 +103,45 @@ def plot(
...
         # Set labels of graph
         if self.fdata.dataset_name is not None:
-            fig.suptitle(self.fdata.dataset_name)
-        ax.set_xlabel("MEI")
-        ax.set_ylabel("MBD")
-        ax.set_xlim([0, 1])
-        ax.set_ylim([
+            self.axScatter.set_title(self.fdata.dataset_name)
+        else:
+            self.axScatter.set_title("Outliergram")
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:135]: Remove the else.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/exploratory/visualization/_baseplot.py (hunk 1 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_baseplot.py:37]: the problem is that i have to use np.array as it allows me to create an empty array (which is needed everytime i do plot)
besides i only need integers and ndarray gives me floats

File: skfda/exploratory/visualization/_outliergram.py (hunk 3 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py:135]: ok

File: skfda/exploratory/visualization/_outliergram.py (hunk 2 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py:110]: ok

File: skfda/exploratory/visualization/_baseplot.py (hunk 1 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_baseplot.py:37]: ok i just realized its type ndarray
my bad, going to correct it :)

File: skfda/exploratory/visualization/_outliergram.py [hunk 4]
```diff
@@ -139,18 +156,91 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
+        The distances can be calculated with function:
+            d_i = a_0 + a_1* mei_i + n^2* a_2* mei_i^2 - mb_i.
+        """
+        distances = []
+        parable = []
+        a_0 = -2 / (self.n * (self.n - 1))
+        a_1 = (2 * (self.n + 1)) / (self.n - 1)
+        a_2 = a_0
+
+        for mbd_item, mei_item in zip(self.mbd, self.mei):
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:190]: Can't this be vectorized?

File: skfda/exploratory/visualization/_outliergram.py [hunk 5]
```diff
@@ -139,18 +156,91 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
+            p_i = (
+                a_0 + a_1 * mei_item + pow(self.n, 2) * a_2 * pow(mei_item, 2)
+            )
+            distances.append(p_i - mbd_item)
+            parable.append(p_i)
+        return distances, parable
+
+    def compute_outliergram(self) -> None:
+        """Compute the parabola under which the outliers lie."""
+        percentile_25 = 25
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:200]: Really? Probably it is more reasonable to ignore the flake8 error.

File: skfda/exploratory/visualization/_outliergram.py [hunk 6]
```diff
@@ -139,18 +156,91 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
+        first_quartile = np.percentile(self.distances, percentile_25)
+        third_quartile = np.percentile(self.distances, percentile_75)
+        iqr = third_quartile - first_quartile
+        self.shifted_parable = self.parable - (third_quartile + iqr)
+
+    def n_samples(self) -> int:
+        """Get the number of instances that will be used for interactivity."""
+        return self.fdata.n_samples
+
+    def set_figure_and_axes(
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:211]: This should not be a public method.

File: skfda/exploratory/visualization/_outliergram.py [hunk 7]
```diff
@@ -139,18 +156,91 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
+
+        for mbd_item, mei_item in zip(self.mbd, self.mei):
+            p_i = (
+                a_0 + a_1 * mei_item + pow(self.n, 2) * a_2 * pow(mei_item, 2)
+            )
+            distances.append(p_i - mbd_item)
+            parable.append(p_i)
+        return distances, parable
+
+    def compute_outliergram(self) -> None:
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:198]: Probably these "compute" methods should not be public.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/exploratory/visualization/_outliergram.py (hunk 4 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py]: true, done

File: skfda/exploratory/visualization/_outliergram.py [hunk 8]
```diff
@@ -50,22 +63,39 @@ class Outliergram:
     def __init__(
         self,
         fdata: FDataGrid,
+        *,
+        chart: Union[Figure, Axes, None] = None,
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:67]: The only reason for having a `chart` parameter is to be positional only. If it does not make sense for it to be positional, maybe it can be removed (and the code will be simpler).

File: skfda/exploratory/visualization/_outliergram.py [hunk 9]
```diff
@@ -139,18 +156,86 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
+            num_functions_above,
             x=self.fdata.grid_points[0],
             axis=1,
         )
 
         integrand /= (interval_len * self.fdata.n_samples)
 
         return integrand.flatten()
+
+    def __compute_distances(self) -> np.ndarray:
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:177]: I prefer protected methods (one underscore) rather than private ones (two underscores) unless there is a very good reason. Private names have a little of "magic" by prepending the class name.

File: skfda/exploratory/visualization/_outliergram.py [hunk 10]
```diff
@@ -174,41 +174,36 @@ def modified_epigraph_index_list(self) -> np.ndarray:
...
         return distances, parable
 
-    def compute_outliergram(self) -> None:
+    def __compute_outliergram(self) -> None:
         """Compute the parabola under which the outliers lie."""
-        percentile_25 = 25
-        percentile_75 = 75
-        first_quartile = np.percentile(self.distances, percentile_25)
-        third_quartile = np.percentile(self.distances, percentile_75)
+        first_quartile = np.percentile(self.distances, 25)
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:197]: Use a `# noqa: WPS432` comment.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/exploratory/visualization/_outliergram.py (hunk 8 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py:67]: i can move it over the *

File: skfda/exploratory/visualization/_outliergram.py [hunk 11]
```diff
@@ -50,22 +63,39 @@ class Outliergram:
...
+        n_cols: Optional[int] = None,
+        **kwargs,
     ) -> None:
+        BasePlot.__init__(self)
         self.fdata = fdata
         self.depth = ModifiedBandDepth()
         self.depth.fit(fdata)
         self.mbd = self.depth(fdata)
         self.mei = self.modified_epigraph_index_list()
+        if self.mbd.size != self.mei.size:
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:80]: I think `len` is more clear than `size` (the latter wouldn't work if the numpy array had more than one dimension.

File: skfda/exploratory/visualization/_outliergram.py [hunk 12]
```diff
@@ -50,22 +63,39 @@ class Outliergram:
...
+            raise ValueError(
+                "The size of mbd and mei should be the same.",
+            )
+        self.n = self.mbd.size
+        distances, parable = self._compute_distances()
+        self.distances = distances
+        mei_ordered = self.mei[:]
+        mei_ordered, parable = (
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:88]: Use `np.argsort` to find the ordered indexes. It should be more efficient.

File: skfda/exploratory/visualization/_outliergram.py [hunk 13]
```diff
@@ -50,22 +63,39 @@ class Outliergram:
...
         self.mbd = self.depth(fdata)
         self.mei = self.modified_epigraph_index_list()
+        if self.mbd.size != self.mei.size:
+            raise ValueError(
+                "The size of mbd and mei should be the same.",
+            )
+        self.n = self.mbd.size
+        distances, parable = self._compute_distances()
+        self.distances = distances
+        mei_ordered = self.mei[:]
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:87]: I think this line is useless.

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/exploratory/visualization/_outliergram.py (hunk 12 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py:87]: but i dont want the indices, i want to order the parable array as the mei, so that i can plot it in order

Comment (vnmabus) [reply skfda/exploratory/visualization/_outliergram.py:87]: Once you have the indices you can use them to index both arrays.

Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py]: ok true, i was testing it wrongly

**Comment by mellamansanchez** [comment]
can this plot be merged, i changed all the mentioned

File: skfda/exploratory/visualization/_outliergram.py [hunk 14]
```diff
@@ -50,22 +63,36 @@ class Outliergram:
...
         self.fdata = fdata
         self.depth = ModifiedBandDepth()
         self.depth.fit(fdata)
         self.mbd = self.depth(fdata)
         self.mei = self.modified_epigraph_index_list()
+        if len(self.mbd) != len(self.mei):
+            raise ValueError(
+                "The size of mbd and mei should be the same.",
+            )
+        self.n = self.mbd.size
```
Comment (vnmabus) [inline skfda/exploratory/visualization/_outliergram.py:84]: Isn't this the same as `n_samples`?

**Review by vnmabus — CHANGES_REQUESTED** [review]

File: skfda/exploratory/visualization/_outliergram.py (hunk 14 above)
Comment (mellamansanchez) [reply skfda/exploratory/visualization/_outliergram.py]: true, just changed it

**Review by vnmabus — APPROVED** [review]
LGTM
